Some technical improvements

hwoarangmy opened the pull request.

hwoarangmy committed:
AI should build rooms in a more generic way

hwoarangmy committed:
Used polymorphism to handle creature actions

fixes #1021

hwoarangmy committed:
Avoid the bool trap in take damage function as requested by @Bertram25

akien-mga commented:
Good to merge as far as I'm concerned. It runs fine and the code looks ok (though I'll admit I did not review df085d7 thoroughly :p)

hwoarangmy commented on source/gamemap/GameMap.cpp:
```diff
@@ -619,6 +619,7 @@ Creature* GameMap::getWorkerToPickupBySeat(Seat* seat)
         bool isClaiming = false;
         bool isDigging = false;
 
+#if 0
```
I forgot to remove this. Note that this prevents merging.

hwoarangmy commented on source/gamemap/GameMap.cpp:
```diff
@@ -736,6 +738,7 @@ Creature* GameMap::getFighterToPickupBySeat(Seat* seat)
         bool isFleeing = false;
         bool isBusy = false;
 
+#if 0
```
same

hwoarangmy committed:
In the takeDamage function, rename damage to absoluteDamage as requested by @akien-mga

hwoarangmy committed:
Remove the "attack" action for creature and use a function instead.

Because we want to check each turn for the closest foe there is no need to remain the last attacked one and, thus, no need for an action

hwoarangmy committed:
Make buildings notify listeners when dead

hwoarangmy committed:
Handle the job room in the creature action rather than at creature level

hwoarangmy committed:
Weak creatures should try to sleep

hwoarangmy committed:
Creatures in arena should not stop fighting because hungry/tired

hwoarangmy committed:
Fixed forgotten stubbed code about choosing best worker/fighter

hwoarangmy committed:
Ensure carried entity is always released

Bertram25 commented:
Ok, I'll review all that tonight.

hwoarangmy committed:
Use different creature actions to handle more efficiently eating

hwoarangmy committed:
Handle the room where creatures eat in the action rather than in the creature class

hwoarangmy committed:
Allow nullptr tile in Tile::displayAsString

hwoarangmy committed:
Handle carried entities destinations on the creature actions instead of the creature class

hwoarangmy commented:
@Bertram25 I'm done with what I wanted to do with the ceature actions ^^
Feel free to review when you can

Bertram25 commented:
Ok :) I'll do that near noon

Bertram25 commented:
I've reviewed the code and it looks good to me. I also had a full game, and everything looks fine so far. (I won but not by far...)
The gameplay has changed a lot, and it's much better. :)
This can be merged IMHO.

Bertram25 commented:
I've merged to move on with this. :)
